src/main/java/com/google/research/bleth/simulator/Simulation.java

```diff
@@ -1,9 +1,399 @@
+         * @return this, to provide chaining.
+         */
+        public SimulationBuilder setBeaconMovementStrategy(IMovementStrategy beaconMovementStrategy) {
+            this.beaconMovementStrategy = beaconMovementStrategy;
+            return this;
+        }
+
+        /**
+         * Set the observers' movement strategy.
+         * @param observerMovementStrategy is the movement strategy for all observers.
+         * @return this, to provide chaining.
+         */
+        public SimulationBuilder setObserverMovementStrategy(IMovementStrategy observerMovementStrategy) {
+            this.observerMovementStrategy = observerMovementStrategy;
+            return this;
+        }
+
+        /**
+         * Set the observers' awakeness strategy factory, used for generating awakeness strategies for all observers.
+         * @param awakenessStrategyFactory is the awakeness strategy factory for all observers.
+         * @return this, to provide chaining.
+         */
+        public SimulationBuilder setAwakenessStrategyFactory(AwakenessStrategyFactory awakenessStrategyFactory) {
+            this.awakenessStrategyFactory = awakenessStrategyFactory;
+            return this;
+        }
+
+        /**
+         * Set the threshold transmission radius.
+         * @param radius is the threshold transmission radius
+         * @return this, to provide chaining.
+         */
+        public SimulationBuilder setRadius(double radius) {
+            this.radius = radius;
+            return this;
+        }
+
+        /**
+         * Validate all simulation builder arguments are legal.
+         */
+        abstract void validateArguments();
+
+        /**
+         * Create and initialize simulation observers using a factory, and store them in observers container.
+         * Initializes the observers movement strategy according to the strategy passed to the builder.
+         * Initializes the observers awakeness strategy according to the strategy factory passed to the builder.
+         */
+        abstract void initializeObservers();
+
+        /**
+         * Create and initialize simulation beacons using a factory, and store them in observers container.
+         * Create simple beacons for a tracing simulation and swapping beacons for a stalking simulation.
+         * Initializes the beacons movement strategy according to the strategy passed to the builder.
+         */
+        abstract void initializeBeacons();
+
+        /**
+         * Construct a simulation object.
+         * @return a new Simulation object constructed with the builder parameters.
+         */
+        public abstract Simulation build();
+    }
+
+    /**
+     * An abstract builder class designed to separate the construction of a simulation from its representation.
+     * Designated for the construction of a simulation based on old simulation state read from the db.
+     */
+    public static abstract class SimulationBuilderFromExisting {
```

Comment:
Do you really need a separate builder class for this case? Can't you just add a factory method to `SimulationBuilder` to create it from existing state?

Reply:
I can think of having a single `Builder` inner class, with both `buildFromScratch()` and `buildFromExisting()` methods. Did I get you right?